connaisseur/keys.py

```diff
@@ -0,0 +1,93 @@
+import base64
+import re
+
+import ecdsa
+import rsa
+
+from connaisseur.exceptions import NoSuchClassError
+
+KMS_REGEX = r"^(awskms|gcpkms|azurekms|hashivault|k8s):\/{2,3}[a-zA-Z0-9_.+\/:-]+$"
+KEYLESS_REGEX = r"^[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+$"
+
+
+class KeyInterface:
+    """
+    Interface form which all keys inherit.
+    """
+
+    def __new__(cls, data: object):
+        instance = super(KeyInterface, cls).__new__(cls)
+        instance.__init__(data)
+        return instance
+
+    def __init__(self, data: object) -> None:
+        self.value = data
+
+    def __str__(self) -> str:
+        return self.value
+
+
+class Key(KeyInterface):
+    """
+    Abstract Key class, containing the public key used for verifying the image signatures.
+    """
+
+    value: object
+
+    def __new__(cls, data: str):
+        try:
+            key_cls, key_data = Key.__get_key_type_cls(data)
+            return key_cls.__new__(key_cls, key_data)
+        except Exception as err:
+            msg = "Cannot discern key type."
+            raise NoSuchClassError(message=msg, key_type=key_cls) from err
+
+    @staticmethod
+    def __get_key_type_cls(data: str):
+        if re.match(KEYLESS_REGEX, data):
+            return KeyLessKey, data
+        elif re.match(KMS_REGEX, data):
+            return KMSKey, data
+        elif key := Key.__is_ecdsa_key(data):
+            return ECDSAKey, key
+        elif key := Key.__is_rsa_key(data):
+            return RSAKey, key
+        return None, None
+
+    @staticmethod
+    def __is_ecdsa_key(data: str):
+        try:
+            return ecdsa.VerifyingKey.from_pem(data)
+        except Exception:
+            return None
+
+    @staticmethod
+    def __is_rsa_key(data: str):
+        try:
+            return rsa.PublicKey.load_pkcs1_openssl_pem(data)
+        except Exception:
+            return None
+
+
+class ECDSAKey(KeyInterface):
+    def __str__(self) -> str:
+        return base64.b64encode(self.value.to_der()).decode("utf-8")
+
+
+class RSAKey(KeyInterface):
+    # reminder when implementing RSA validation:
+    # ["--key", "/dev/stdin", self.value.save_pkcs1()]
+
+    def __str__(self) -> str:
+        return base64.b64encode(self.value.save_pkcs1("DER")).decode("utf-8")
+
+
+class KMSKey(KeyInterface):
+    pass
+
+
+class KeyLessKey(KeyInterface):
+    # reminder when implementing Keyless validation:
```

Comment:
While this is more or less guaranteed to be cosign specific, the key still shouldn't hold this info I'd argue